server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -349,6 +363,8 @@ public boolean syncUpTo(long offset) throws IOException {
                         try {
                             outputStream.flush();
                             checkpointToSync = getCheckpoint();
+                            flushedSequenceNumbers = nonFsyncedSequenceNumbers;
```

Comment:
I think we can get the processed checkpoint here and then use it as the persisted checkpoint after we fsync (line 389). With this way, we can avoid the extra work in LocalCheckpointTracker and we can also remove `nonFsyncedSequenceNumbers`.

Reply:
I have been prototyping this in 54e492a, see update to this PR. It requires another change though which I like a little less, and that's that we can't have a `markAsPersisted` method anymore. This means that on a recovery from the translog, we need to manually advance the persisted local checkpoint at the end as the translog writer will not sync (as nothing needs to be done). Also this assumes that no other ongoing indexing operations are happening in the system (which fortunately we don't have right now, but it is easy to break this), as manually advancing the persisted checkpoint is dangerous to do.

Reply:
Thank you for giving this a try :). I agree that advancing the persisted checkpoint is dangerous and we should not do it. How about adding `peristed_checkpoint != processed_checkpoint`  condition to [syncNeeded](https://github.com/elastic/elasticsearch/blob/079e012fdea68ea0a7daae078359495047e9c407/server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java#L263) method. However, I am fine if we decide to go with the previous solution.

Reply:
I've done something quite similar to your suggestion in 798a05e6d83, which makes this all a bit safer. I'm not sure yet how to test the changes to the `syncUpTo` method though, any ideas there are welcome.